docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -313,35 +559,125 @@ type LocalObjectTemplate struct {
       Replicas *int `json:"replicas,omitempty"`
     }
     ```
+1.  The `VariableTopology` object represents an instance of a variable.
+    ```golang
+    // VariableTopology can be used to customize the ClusterClass through
+    // patches. It must comply to the corresponding
+    // VariableDefinitionClass defined in the ClusterClass.
+    type VariableTopology struct {
+      // Name of the variable.
+      Name string `json:"name"`
+
+      // Value of the variable.
+      // Note: the value will be validated against the schema
+      // of the corresponding VariableDefinitionClass from the ClusterClass.
+      Value apiextensionsv1.JSON `json:"value"`
+    }
+    ```
+
+#### Builtin variables
+
+It’s also possible to use so-called builtin variables in addition to user-defined variables. The following builtin variables are available:
+- `cluster.{name,namespace,labels,annotations}`
+- `cluster.topology.{version,class}`
+- `cluster.topology.controlPlane.{labels,annotations,replicas,version}`
+    - **Note**: these variables are only available when patching control plane or control plane machine templates.
+- `cluster.topology.machineDeployment.current.{labels,annotations,replicas,version,class}`
+    - **Note**: these variables are only available when patching MachineDeployment templates and contain the values 
+      of the current `MachineDeploymentTopology`.
+
+Builtin variables are available under the `builtin.` prefix. Some examples:
+- Usage of `cluster.name` via variable:
+   ```yaml
+   jsonPatches:
+   - op: replace
+     path: “/spec/template/spec/kubeadmConfigSpec/clusterConfiguration/controllerManager/extraArgs/cluster-name”
+     valueFrom:
+       variable: builtin.cluster.name
+   ```
+- Usage of `cluster.name` via template:
+   ```yaml
+   jsonPatches:
+   - op: replace
+     path: “/spec/template/spec/kubeadmConfigSpec/files”
+     valueFrom:
+       template: |
+         - contentFrom:
+             secret
+               key: control-plane-azure.json
+               name: {{ .builtin.cluster.name }}-control-plane-azure-json
+   ```
 
 #### Validations
+
 ##### ClusterClass
 - For object creation:
   - (defaulting) if namespace field is empty for a reference, default it to `metadata.Namespace`
   - all the reference must be in the same namespace of `metadata.Namespace`
   - `spec.workers.machineDeployments[i].class` field must be unique within a ClusterClass.
+  - `VariableDefinitionClass`:
+    - names must be unique, not empty and not equal to `builtin`
+    - schemas must be valid
+  - `PatchClass`:
+    - names must be unique and not empty
+  - `PatchDefinition`:
+    - selector must have valid field values and match at least one template
+    - jsonPatches:
```

Comment:
The only problem I see with jsonPatches is the lack of list iteration in JSONPatch. AFAIK it is not possible to change e.g. every item in a list in a single JSONPatch. The length of the list must be known in order to do it.

I don't think that's a blocker as JSONPatch is widely used but something to be aware of (if any usecases for ClusterClass include patching multiple elements in lists).

Reply:
agreed this is not a blocker, also this https://github.com/kubernetes-sigs/cluster-api/pull/5212#discussion_r704184935 could provide a powerful iteration